src/node_large_page.cc

```diff
@@ -0,0 +1,233 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
+// THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES
+// OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+// ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE
+// OR OTHER DEALINGS IN THE SOFTWARE.
+//
+// SPDX-License-Identifier: MIT
+
+#include <errno.h>
+#include <fcntl.h>  // _O_RDWR
+#include <limits.h>  // PATH_MAX
+#include <locale.h>
+#include <signal.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include <string.h>
+#include <sys/types.h>
+#include <sys/mman.h>
+
+#include <string>
+#include <fstream>
+#include <iostream>
+
+// The functions in this file map the text segment of node into 2M pages.
+// The algorithm is quite simple
+// 1. Find the text region of node binary in memory
+// 2. Move the text region to large pages
+
+extern char __executable_start;
+extern char __etext;
+extern char __nodetext;
+
+namespace node {
+namespace largepages {
+#define ALIGN(x, a)           (((x) + (a) - 1) & ~((a) - 1))
+#define PAGE_ALIGN_UP(x, a)   ALIGN(x, a)
+#define PAGE_ALIGN_DOWN(x, a) ((x) & ~((a) - 1))
+
+struct TextRegion {
+  void * from;
+  void * to;
+  int    totalHugePages;
+  int64_t  offset;
+  bool   found_text_region;
+};
+
+    static void printSystemError(int error) {
+      fprintf(stderr, "Hugepages WARNING: %s\n", strerror(error));
+      return;
+    }
+
+    static struct TextRegion find_node_text_region() {
+      FILE *f;
+      int64_t  start, end, offset, inode;
+      char perm[5], dev[6], name[256];
+      int ret;
+      const size_t hugePageSize = 2L * 1024 * 1024;
+      struct TextRegion nregion;
+
+      nregion.found_text_region = false;
+
+      f = fopen("/proc/self/maps", "r");
+      ret = fscanf(f, "%lx-%lx %4s %lx %5s %ld %s\n",
+                   &start, &end, perm, &offset, dev, &inode, name);
+      fclose(f);
+
+      if (ret == 7 &&
+          perm[0] == 'r' && perm[1] == '-' && perm[2] == 'x') {
+        // Checking if the region is from node binary and executable
+        start = (unsigned int64_t) &__nodetext;
+        char *from = reinterpret_cast<char *>PAGE_ALIGN_UP(start, hugePageSize);
+        char *to = reinterpret_cast<char *>PAGE_ALIGN_DOWN(end, hugePageSize);
+
+        if (from < to) {
+          size_t size = (intptr_t)to - (intptr_t)from;
+          nregion.found_text_region = true;
+          nregion.from = from;
+          nregion.to = to;
+          nregion.offset = offset;
+          nregion.totalHugePages = size/hugePageSize;
+          return nregion;
+        }
+      }
+
+      return nregion;
+    }
+
+    static bool isTransparentHugePagesEnabled() {
+       std::ifstream ifs;
+
+       ifs.open("/sys/kernel/mm/transparent_hugepage/enabled");
+       if (!ifs) {
+         fprintf(stderr, "WARNING: Couldn't check hugepages support\n");
+         return false;
```

Comment:
can we follow a consistent return value convention? (returning null vs. -1 vs errno) I suggest look arounf in other APIs for the prevailing convention.

Reply:
Sure